nlmod/mfoutput.py

```diff
@@ -42,31 +50,42 @@ def _get_output_da(reader_func, ds=None, gwf_or_gwt=None, fname=None, **kwargs):
         hdry = -1e30
         hnoflo = 1e30
 
-    # check whether out_obj is BudgetFile or HeadFile based on passed kwargs
-    if isinstance(out_obj, CellBudgetFile):
-        arr = out_obj.get_data(**kwargs)
-    elif isinstance(out_obj, HeadFile):
-        arr = out_obj.get_alldata(**kwargs)
+    if "kstpkper" in kwargs:
+        kstpkper = kwargs.pop("kstpkper")
     else:
-        raise TypeError(f"Don't know how to deal with {type(out_obj)}!")
+        kstpkper = out_obj.get_kstpkper()
 
-    if isinstance(arr, list):
-        arr = np.stack(arr)
+    result = []
+    for ki in kstpkper:
+        d = dask.delayed(out_obj.get_data)(kstpkper=ki, **kwargs)
+        arr = dask.array.from_delayed(d, shape=out_obj.mg.shape, dtype=float)
+        result.append(arr)
 
-    arr[arr == hdry] = np.nan
-    arr[arr == hnoflo] = np.nan
+    stacked_arr = dask.array.stack(result)
 
     if gwf_or_gwt is not None:
         gridtype = gwf_or_gwt.modelgrid.grid_type
     else:
         gridtype = ds.gridtype
 
     if gridtype == "vertex":
+        layers = np.arange(stacked_arr.shape[1])
```

Comment:
If a dataset is given I think it is better to use the layers from the dataset. Otherwise you might overwrite the layers dimension of the dataset.